File: csharp/ql/test/query-tests/Language Abuse/SimplifyBoolExpr/SimplifyBoolExpr.cs

class Test
{
    void Fn()
    {
        bool b = false;
        int x = 0, y = 0;

        if (b == true) ; // $ Alert
        if (b == false) ; // $ Alert
        if (true == b) ; // $ Alert
        if (false == b) ; // $ Alert
        if (b != true) ; // $ Alert
        if (b != false) ; // $ Alert
        if (true != b) ; // $ Alert
        if (false != b) ; // $ Alert
        if (b && true) ; // $ Alert
        if (b && false) ; // $ Alert
        if (true && b) ; // $ Alert
        if (false && b) ; // $ Alert
        if (b || true) ; // $ Alert
        if (b || false) ; // $ Alert
        if (true || b) ; // $ Alert
        if (false || b) ; // $ Alert
        if (!(x == y)) ; // $ Alert
        if (!(x != y)) ; // $ Alert
        if (!(x < y)) ; // $ Alert
        if (!(x <= y)) ; // $ Alert
        if (!(x >= y)) ; // $ Alert
        if (!(x > y)) ; // $ Alert
        if (b ? true : false) ; // $ Alert
        if (b ? true : true) ; // $ Alert
        if (b ? false : true) ; // $ Alert
        if (b ? true : true) ; // $ Alert
        if (b ? b : false) ; // $ Alert
        if (b ? b : true) ; // $ Alert
        if (b ? false : b) ; // $ Alert
        if (b ? true : b) ; // $ Alert

        // BAD
        if (true ? b : b) ;
        if (false ? b : b) ;
        if (true == true) ;
        if (true != false) ;
        if (true && true) ;
        if (true || false) ;
    }
}

File: java/ql/test/query-tests/SimplifyBoolExpr/SimplifyBoolExpr.java
class Test {
  void f(boolean x, boolean y, Boolean a, Boolean b) {
    boolean w;
    w = a == false; // $ Alert
    w = x != true; // $ Alert
    w = a ? false : b; // $ Alert
    w = a ? true : false; // $ Alert
    w = x ? y : true; // $ Alert
  }
  void g(int x, int y) {
    boolean w;
    w = !(x > y); // $ Alert
    w = !(x != y); // $ Alert
  }
  public Boolean getBool(int i) {
    if (i > 2)
      return i == 3 ? true : null; // ok; expression can't be simplified
    return i == 1 ? false : null; // ok; expression can't be simplified
  }
  public Boolean getBoolNPE(int i) {
    if (i > 2)
      return i == 3 ? true : ((Boolean)null); // $ Alert // should be reported; both this and the simplified version have equal NPE behavior
    return i == 1 ? false : ((Boolean)null); // $ Alert // should be reported; both this and the simplified version have equal NPE behavior
  }
}
